packages/fetchai/connections/soef/connection.py

```diff
@@ -91,106 +147,349 @@ def __init__(
+
+        for constraint in equality_constraints:
+            if constraint.attribute_name not in PERSONALITY_PIECES:
+                continue
+            filters.append(
+                constraint.attribute_name + "," + constraint.constraint_type.value
+            )
+        if not filters:  # pragma: nocover
+            return {}
+        return {"ppfilter": filters}
+
+    @staticmethod
+    def _construct_service_key_filter_params(
+        equality_constraints: List[Constraint],
+    ) -> Dict[str, List[str]]:
+        """
+        Construct a dictionary of service keys filters.
+
+        :param equality_constraints: list of equality constraints
+        :return: bool
+        """
+        filters = []
+
+        for constraint in equality_constraints:
+            if constraint.attribute_name in PERSONALITY_PIECES:
+                continue
+            filters.append(
+                constraint.attribute_name + "," + constraint.constraint_type.value
+            )
+        if not filters:  # pragma: nocover
+            return {}
+        return {"skfilter": filters}
+
+    def _check_protocol_valid(self, envelope: Envelope) -> None:
+        """
+        Check protocol is supported and raises ValueError if not.
+
+        :param envelope: envelope to check protocol of
         :return: None
         """
-        if self.excluded_protocols is not None:
-            if envelope.protocol_id in self.excluded_protocols:
-                logger.error(
-                    "This envelope cannot be sent with the soef connection: protocol_id={}".format(
-                        envelope.protocol_id
-                    )
+        is_in_excluded = envelope.protocol_id in (self.excluded_protocols or [])
+        is_in_restricted = not self.restricted_to_protocols or envelope.protocol_id in (
+            self.restricted_to_protocols or []
+        )
+
+        if is_in_excluded or not is_in_restricted:
+            logger.error(
+                "This envelope cannot be sent with the soef connection: protocol_id={}".format(
+                    envelope.protocol_id
                 )
-                raise ValueError("Cannot send message.")
-        if envelope.protocol_id in self.restricted_to_protocols:
-            assert (
-                envelope.protocol_id == OefSearchMessage.protocol_id
-            ), "Invalid protocol id passed check."
-            self.process_envelope(envelope)
-        else:
+            )
             raise ValueError(
                 "Cannot send message, invalid protocol: {}".format(envelope.protocol_id)
             )
 
-    def process_envelope(self, envelope: Envelope) -> None:
+    async def send(self, envelope: Envelope) -> None:
+        """
+        Send message handler.
+
+        :param envelope: the envelope.
+        :return: None
+        """
+        self._check_protocol_valid(envelope)
+        await self.process_envelope(envelope)
+
+    async def _request_text(self, *args, **kwargs) -> str:
```

Comment:
Where is this used?

Reply:
it's a primary code to perform every http request and return text, its used by generic oef request and in agent register